docs/architecture/adr-062-p2p-architecture.md

```diff
@@ -0,0 +1,531 @@
+type StreamID uint8
+
+// Stream represents a single logical IO stream within a connection.
+type Stream interface {
+    io.Reader // Read([]byte) (int, error)
+    io.Writer // Write([]byte) (int, error)
+    io.Closer // Close() error
+}
+```
+
+### Peers
+
+Peers are other Tendermint network nodes. Each peer is identified by a unique `PeerID`, and has a set of `PeerAddress` addresses expressed as URLs that they can be reached at. Examples of peer addresses might be e.g.:
+
+* `mconn://b10c@host.domain.com:25567/path`
+* `unix:///var/run/tendermint/peer.sock`
+* `memory:testpeer`
+
+Addresses are resolved into one or more transport endpoints, e.g. by resolving DNS hostnames into IP addresses (which should be refreshed periodically). Peers should always be expressed as address URLs, and never as endpoints which are a lower-level construct.
+
+```go
+// PeerID is a unique peer ID, generally expressed in hex form.
+type PeerID []byte
+
+// PeerAddress is a peer address URL. The User field, if set, gives the
+// hex-encoded remote PeerID, which should be verified with the remote peer's
+// public key as returned by the connection.
+type PeerAddress url.URL
+
+// Resolve resolves a PeerAddress into a set of Endpoints, typically by
+// expanding out any DNS names given in Host to IP addresses. Field mapping:
+//
+//   Scheme → Endpoint.Protocol
+//   Host   → Endpoint.IP
+//   Port   → Endpoint.Port
+//   Path+Query+Fragment,Opaque → Endpoint.Path
+//
+func (a PeerAddress) Resolve(ctx context.Context) []Endpoint { return nil }
+```
+
+The P2P stack needs to track a lot of internal information about peers, such as endpoints, status, priorities, and so on. This is done in an internal `peer` struct, which should not be exposed outside of the `p2p` package (e.g. to reactors) in order to avoid race conditions and lock contention - other packages should use `PeerID`.
+
+The `peer` struct might look like the following, but is intentionally underspecified and will depend on implementation requirements (for example, it will almost certainly have to track statistics about connection failures and retries):
+
+```go
+// peer tracks internal status information about a peer.
+type peer struct {
+    ID        PeerID
+    Status    PeerStatus
+    Priority  PeerPriority
+    Endpoints map[PeerAddress][]Endpoint // Resolved endpoints by address.
+}
+
+// PeerStatus specifies peer statuses.
+type PeerStatus string
+
+const (
+    PeerStatusNew     = "new"     // New peer which we haven't tried to contact yet.
+    PeerStatusUp      = "up"      // Peer which we have an active connection to.
+    PeerStatusDown    = "down"    // Peer which we're temporarily disconnected from.
+    PeerStatusRemoved = "removed" // Peer which has been removed.
+    PeerStatusBanned  = "banned"  // Peer which is banned for misbehavior.
+)
+
+// PeerPriority specifies peer priorities.
+type PeerPriority int
+
+const (
+    PeerPriorityNormal PeerPriority = iota + 1
```

Comment:
will this lead to `if` conditions in the code? maybe it's better to make PeerPriority a simple number from 0 to 10 (we can even expose it via config - e.g. `unix:///var/run/tendermint/peer.sock||P:10`; just an idea)

Reply:
Yes, `PeerPriority` is meant to be a number. This example is just a simple enumeration of them, but we could use any sort of scheme we want, including multiple weighted conditions. This needs to be specced out when we design the router, since that's the component which will prioritize and maintain peer connections.